<a href="https://colab.research.google.com/github/vard-uhi/Retention-Analysis/blob/master/Human_Resources_Retention.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Retention Study - Figuring Out Which Employees May Quit**

#### Loading our HR Database

In [14]:
import numpy as np
import pandas as pd
import pandas.testing as tm
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [15]:
#load the data
from google.colab import files
data_to_load = files.upload()

Saving hr_data.csv to hr_data (1).csv


In [16]:
import io
raw_data = pd.read_csv(io.BytesIO(data_to_load['hr_data.csv']))
raw_data.head()

employee_id  number_project  ...  department  salary
0         1003               2  ...       sales     low
1         1005               5  ...       sales  medium
2         1486               7  ...       sales  medium
3         1038               5  ...       sales     low
4         1057               2  ...       sales     low

[5 rows x 9 columns]

In [17]:
#copying our raw Dataframe, in order to performe manipulations
hr_df = raw_data.copy()

In [18]:
#general overview of the DataFrame
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   employee_id            14999 non-null  int64 
 1   number_project         14999 non-null  int64 
 2   average_montly_hours   14999 non-null  int64 
 3   time_spend_company     14999 non-null  int64 
 4   Work_accident          14999 non-null  int64 
 5   left                   14999 non-null  int64 
 6   promotion_last_5years  14999 non-null  int64 
 7   department             14999 non-null  object
 8   salary                 14999 non-null  object
dtypes: int64(7), object(2)
memory usage: 1.0+ MB


In [29]:
hr_df[hr_df.isnull().any(axis=1)]

Empty DataFrame
Columns: [employee_id, number_project, average_montly_hours, time_spend_company, Work_accident, left, promotion_last_5years, department, salary]
Index: []


*   We have 2 categorical variables 'department', 'salary'
*   No missing values  
*   Shape of the Dataframe is 14999 rows and 9 columns

#### Loading Evaluation and Employee Satisfaction Data

In [19]:
#load the data
from google.colab import files
data_to_load = files.upload()

Saving employee_satisfaction_evaluation.xlsx to employee_satisfaction_evaluation (1).xlsx


In [20]:
import io
raw_data2 = pd.read_excel(io.BytesIO(data_to_load['employee_satisfaction_evaluation.xlsx']))
raw_data2.head()

EMPLOYEE #  satisfaction_level  last_evaluation
0        1003                0.38             0.53
1        1005                0.80             0.86
2        1486                0.11             0.88
3        1038                0.72             0.87
4        1057                0.37             0.52

In [21]:
emp_satis_eval = raw_data2.copy()

In [22]:
emp_satis_eval.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 3 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   EMPLOYEE #          14999 non-null  int64  
 1   satisfaction_level  14972 non-null  float64
 2   last_evaluation     14972 non-null  float64
dtypes: float64(2), int64(1)
memory usage: 351.7 KB


* We have missing vales.
* No categorical data.
* Shape is 14999 rows and 3 columns

Next we will merge 2 datasets to perform data manipluation, analysis and prediction.

#### Join Tables

In [23]:
main_df = hr_df.set_index('employee_id').join(emp_satis_eval.set_index('EMPLOYEE #'))
main_df = main_df.reset_index()
main_df.head()

employee_id  number_project  ...  satisfaction_level  last_evaluation
0         1003               2  ...                0.38             0.53
1         1005               5  ...                0.80             0.86
2         1486               7  ...                0.11             0.88
3         1038               5  ...                0.72             0.87
4         1057               2  ...                0.37             0.52

[5 rows x 11 columns]

In [25]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            14999 non-null  int64  
 1   number_project         14999 non-null  int64  
 2   average_montly_hours   14999 non-null  int64  
 3   time_spend_company     14999 non-null  int64  
 4   Work_accident          14999 non-null  int64  
 5   left                   14999 non-null  int64  
 6   promotion_last_5years  14999 non-null  int64  
 7   department             14999 non-null  object 
 8   salary                 14999 non-null  object 
 9   satisfaction_level     14972 non-null  float64
 10  last_evaluation        14972 non-null  float64
dtypes: float64(2), int64(7), object(2)
memory usage: 1.3+ MB


In [26]:
#descriptive statistics of all variables, including categorical
main_df.describe(include='all')

employee_id  number_project  ...  satisfaction_level  last_evaluation
count   14999.000000    14999.000000  ...        14972.000000     14972.000000
unique           NaN             NaN  ...                 NaN              NaN
top              NaN             NaN  ...                 NaN              NaN
freq             NaN             NaN  ...                 NaN              NaN
mean    45424.627575        3.803054  ...            0.612830         0.716125
std     25915.900127        1.232592  ...            0.248714         0.171138
min      1003.000000        2.000000  ...            0.090000         0.360000
25%     22872.500000        3.000000  ...            0.440000         0.560000
50%     45448.000000        4.000000  ...            0.640000         0.720000
75%     67480.500000        5.000000  ...            0.820000         0.870000
max     99815.000000        7.000000  ...            1.000000         1.000000

[11 rows x 11 columns]

####Dealing with missing values

In [33]:
#list of all rows with missing vales
main_df[main_df.isnull().any(axis=1)]

employee_id  number_project  ...  satisfaction_level  last_evaluation
18            3794               2  ...                 NaN              NaN
19            1140               5  ...                 NaN             0.89
33            1230               2  ...                 NaN              NaN
53            1340               2  ...                 NaN              NaN
72           22316               2  ...                 NaN              NaN
92            1581               2  ...                 NaN              NaN
107          17376               2  ...                 NaN              NaN
120           1739               4  ...                 NaN              NaN
137           1847               2  ...                 NaN              NaN
175          32923               4  ...                 NaN              NaN
191           2160               4  ...                0.92              NaN
352           3150               4  ...                 NaN             0.91
376           3250               4  ...                0.56              NaN
402           3405               5  ...                 NaN             1.00
427          78130               3  ...                 NaN             0.60
442           3635               5  ...                0.77              NaN
468           3755               5  ...                 NaN             0.92
543           4150               5  ...                0.89              NaN
892          43615               4  ...                0.74              NaN
1588         42185               5  ...                 NaN             0.81
1934         11895               4  ...                0.78              NaN
2343         14170               3  ...                 NaN             0.97
2743         16445               5  ...                0.62              NaN
3170         18980               5  ...                 NaN              NaN
3609         21580               3  ...                0.95              NaN
3776         22555               4  ...                 NaN             0.68
4122         24505               3  ...                0.86              NaN
4740         27950               3  ...                 NaN             0.95
5028         29640               4  ...                0.64              NaN
6453         38090               5  ...                 NaN             0.78
7005         41535               4  ...                0.82              NaN
7516         44460               4  ...                 NaN             0.97
8630         50960               4  ...                0.90              NaN
9455         55770               4  ...                 NaN             0.88
9901         58630               5  ...                0.92              NaN
10647        62595               4  ...                 NaN             0.69
10962        64350               5  ...                0.65              NaN
11575        20215               3  ...                 NaN              NaN
11967        70005               3  ...                0.82              NaN
12422        72475               5  ...                 NaN             1.00
12853        74880               3  ...                 NaN             0.70
13482        78780               3  ...                0.58              NaN
13925        81315               3  ...                 NaN             0.75

[43 rows x 11 columns]

In [32]:
#counting the numer of missing rows
main_df[main_df.isnull().any(axis=1)].count()

employee_id              43
number_project           43
average_montly_hours     43
time_spend_company       43
Work_accident            43
left                     43
promotion_last_5years    43
department               43
salary                   43
satisfaction_level       16
last_evaluation          16
dtype: int64

we see that we have 43 rows with missing vales and they come from 'satisfication level' and 'last evaluation' columns.

To deal with missing vales we will fill in the missing blanks with the aveage values.